### PR TITLE
#1 initialize project structure

### PR DESCRIPTION
**Pull request by UdjinSkobelev**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.5.6</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.social</groupId>
+            <artifactId>spring-social-facebook</artifactId>
```
Comment (Max-Levitskiy) [inline pom.xml]: Why do you have facebook here? :)

File: src/main/java/org/webtree/social/stackexchange/api/StackExchange.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.webtree.social.stackexchange.api;
+
+/**
+ * Created by Udjin Skobelev on 27.09.2018.
+ */
+public interface StackExchange extends Api{
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/StackExchange.java]: You shouldn't have here methods from Api. Use composition vs inheritance.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    public void handleError(ClientHttpResponse response) throws IOException {
+        StackExchangeError error = extractErrorFromResponse(response);
+        handleStackExchangeError(response.getStatusCode(), error);
+    }
+
+    void handleStackExchangeError(HttpStatus statusCode, StackExchangeError error) {
+        if (error != null && error.getCode() != null) {
+            int code = error.getCode();
+            switch (code) {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: Transform to map

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+                logger.debug("StackExchange error: ");
+                logger.debug("   CODE     : {}", jsonNode.get("error_id"));
+                logger.debug("   MESSAGE  : {}", jsonNode.get("error_message"));
+                logger.debug("   NAME     : {}", jsonNode.get("error_name"));
+
+                return error;
+            }
+
+        } catch (JsonParseException var13) {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: don't use var13 name. it's awful. :)

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            return null;
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: why do you return null?

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+
+/**
+ * Created by Udjin Skobelev on 27.09.2018.
+ */
+
+public class StackExchangeTemplate implements ApiBinding, StackExchange {
+
+    private final String accessToken;
+    private String apiVersion = "2.2";
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: extract in properties file

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        return null;
+    }
+
+    private String readResponseJson(ClientHttpResponse response) throws IOException {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: Why do you do this? As I remember, Jackson can read from input stream for you.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    public UserOperations userOperations() {
+        return userOperations;
+    }
+
+    public SiteOperations siteOperations() {
+        return siteOperations;
+    }
+
+    public String getBaseApiUrl() {
+        return "https://api.stackexchange.com/" + apiVersion + "/";
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Don't build this string each time. You can create it in constructor and use as a field in class.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            public Type getType() {
+                return new ParameterizedTypeImpl((ParameterizedType) super.getType(), new Type[]{type});
+            }
+        }).getBody();
+    }
+
+    @Override
+    public <T> ResponseWrapper<T> fetchObject(String methodName, Class<T> itemsType, MultiValueMap<String, String> queryParams) {
+        URI uri = URIBuilder.fromUri(getBaseApiUrl() + methodName).queryParams(queryParams).build();
+        return getRestTemplate().exchange(uri, HttpMethod.GET, null, new ParameterizedTypeReference<ResponseWrapper<T>>() {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Why do you call getter for restTemplate?

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+    public String getBaseApiUrl() {
+        return "https://api.stackexchange.com/" + apiVersion + "/";
+    }
+
+    RestTemplate getRestTemplate() {
+        return restTemplate;
+    }
+
+    private FormHttpMessageConverter getFormMessageConverter() {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: get -> create

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        partConverters.add(new ByteArrayHttpMessageConverter());
+        StringHttpMessageConverter stringHttpMessageConverter = new StringHttpMessageConverter(Charset.forName("UTF-8"));
+        stringHttpMessageConverter.setWriteAcceptCharset(false);
+        partConverters.add(stringHttpMessageConverter);
+        partConverters.add(new ResourceHttpMessageConverter());
+        converter.setPartConverters(partConverters);
+        return converter;
+    }
+
+    private MappingJackson2HttpMessageConverter getJsonMessageConverter() {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: get -> create

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    RestTemplate getRestTemplate() {
+        return restTemplate;
+    }
+
+    private FormHttpMessageConverter getFormMessageConverter() {
+        FormHttpMessageConverter converter = new FormHttpMessageConverter();
+        converter.setCharset(Charset.forName("UTF-8"));
+        List<HttpMessageConverter<?>> partConverters = new ArrayList();
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: extract partConverter as method

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        List<ClientHttpRequestInterceptor> interceptors = Arrays.asList(
+                new BasicHttpRequestInterceptor("key", key),
+                new BasicHttpRequestInterceptor("access_token", accessToken));
+        client.setInterceptors(interceptors);
+        client.setErrorHandler(new StackExchangeErrorHandler());
+        return client;
+    }
+
+    private RestTemplate createRestTemplateWithCulledMessageConverters() {
+        List messageConverters = getMessageConverters();
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: You've lost generic here. :)

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        RestTemplate client = new RestTemplate(messageConverters);
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Don't create useless variables. Just return.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        return client;
+    }
+
+    private List<HttpMessageConverter<?>> getMessageConverters() {
+        List<HttpMessageConverter<?>> messageConverters = new ArrayList();
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: ArrayList<>

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        ByteArrayHttpMessageConverter converter = new ByteArrayHttpMessageConverter();
+        converter.setSupportedMediaTypes(Arrays.asList(MediaType.IMAGE_JPEG, MediaType.IMAGE_GIF, MediaType.IMAGE_PNG));
+        return converter;
+    }
+
+    private void setRequestFactory(ClientHttpRequestFactory requestFactory) {
+        restTemplate.setRequestFactory(requestFactory);
+    }
+
+    private void initSubApis() {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: What is SubApis? it's bad name. Create operations is better. Delegate creation or even just providing this operations for somebody else is much better.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        this.userOperations = new UserTemplate(this, getRestTemplate());
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: You definitely shouldn't create cyclic dependency here. The problem came from StackExchange extends Api

File: src/main/java/org/webtree/social/stackexchange/api/impl/UserTemplate.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.List;
+
+/**
+ * Created by Udjin Skobelev on 27.09.2018.
+ */
+
+public class UserTemplate implements UserOperations {
+    private final Api api;
+    private final RestTemplate restTemplate;
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/UserTemplate.java]: You don't use restTemplate here. Remove it.

File: src/main/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverter.java [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Created by Udjin Skobelev on 07.10.2018.
+ */
+
+public class UnixTimestampToMillisecondsConverter extends JsonDeserializer<Date> {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverter.java]: This name is lying. It converts timestamp in milliseconds to date.

File: src/main/java/org/webtree/social/stackexchange/domain/NetworkUser.java [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Date;
+
+/**
+ * Created by Udjin Skobelev on 30.09.2018.
+ */
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NetworkUser {
+
+    @JsonProperty("question_count")
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/domain/NetworkUser.java]: Maybe you can create some default converter which will automatically map underscores to camel case. I think it's not a new problem. :)

File: src/main/java/org/webtree/social/stackexchange/domain/User.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    public String getProfileImage() {
+        return profileImage;
+    }
+
+    public String getDisplayName() {
+        return displayName;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/domain/User.java:124]: Replace all equals-hashcodes to other template based on java.util.Objects (in idea - last option on generation), It produce less boilerplate code.

File: src/test/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandlerTest.java [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+import java.io.IOException;
+
+import static org.webtree.social.stackexchange.api.ErrorCodes.*;
+
+
+/**
+ * Created by Udjin Skobelev on 05.10.2018.
+ */
+
+public class StackExchangeErrorHandlerTest {
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandlerTest.java:27]: Seems this test should be refactored to parameterized test. Maybe it's better to migrate all tests to junit5?

File: src/test/java/org/webtree/social/stackexchange/api/impl/UserTemplateTest.java [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        List<NetworkUser> users = Arrays.asList(fromStack, fromMath);
+        responseWrapper = new ResponseWrapper<>(users, true, 1, 2);
+
+        server
+                .expect(requestTo(stackExchange.getBaseApiUrl() + "me/associated?" + tokenQueryParams))
+                .andRespond(withSuccess(objectMapper.writeValueAsString(responseWrapper), MediaType.APPLICATION_JSON));
+
+        List<NetworkUser> usersFromApi = stackExchange.userOperations().getUserAssociatedAccounts();
+        assertThat(usersFromApi).containsExactlyInAnyOrder(fromStack, fromMath);
+        server.verify();
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/UserTemplateTest.java]: extract to tearDown method

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+public abstract class AbstractApiIntegrationTest {
+    static String TOKEN;
+    static String APP_KEY;
+    StackExchange stackExchange;
+
+    @Before
+    public void setUp() throws Exception {
+        TOKEN =  System.getenv("stackexchange.token");
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java]: stackexchange.token -> stackexchange_token

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+        APP_KEY = System.getenv("stackexchange.key");
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java]: stackexchange.key -> stackexchange_key

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/SiteTemplateIntegrationTest.java [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+        assertThat(sites).isNotEmpty();
+    }
+
+    @Test
+    public void shouldHasSelectedSiteInResponse() {
+        List<Site> sites = stackExchange.siteOperations().getActualSites();
+        Optional<Site> selectedSite = sites.stream().filter((site) -> site.getApiSiteParameter().equals(SELECTED_SITE)).findFirst();
+        assertThat(selectedSite.isPresent()).isTrue();
+        assertThat(selectedSite.get().getApiSiteParameter()).isNotNull();
+        assertThat(selectedSite.get().getName()).isNotNull();
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/integration/SiteTemplateIntegrationTest.java]: You can check real site name here. There is a small chance they change it. :) 
Or at least not empty string.

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/UserTemplateIntegrationTest.java [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+        assertThat(users.get(0).getSiteName()).isNotNull();
+        assertThat(users.get(0).getAccountId()).isNotNull();
+    }
+
+    @Test
+    public void shouldReturnUserBySiteName() {
+        List<User> users = stackExchange.userOperations().getUserProfileAssociatedWithSite(SITE_API_NAME);
+        assertThat(users).isNotEmpty();
+        assertThat(users.get(0).getReputation()).isNotNull().isPositive();
+        assertThat(users.get(0).getAccountId()).isNotNull();
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/api/impl/integration/UserTemplateIntegrationTest.java]: Again, not empty string would be better.

File: src/test/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverterTest.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+public class UnixTimestampToMillisecondsConverterTest {
+    private final static String UNIX_TIME_IN_SECONDS = "1217462400";
+
+    @Mock
+    private JsonParser parser;
+    private UnixTimestampToMillisecondsConverter converter;
+
+    @Test
+    public void name() throws IOException {
+        Date dateAfterConverting = new Date(TimeUnit.SECONDS.toMillis(Long.valueOf(UNIX_TIME_IN_SECONDS)));
```
Comment (Max-Levitskiy) [inline src/test/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverterTest.java]: Don't create copy-paste between test and production code. If there is a problem in logic your test won't find it. 
Here you should use a pair of unix and date from human readable format. The first one should be returned from mock, the second one should be used in accertion.

**Review by Max-Levitskiy — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (UdjinSkobelev) [reply pom.xml]: Mistake, deleted/

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 13 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Now not.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 14 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 15 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/UserTemplate.java (hunk 18 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/UserTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 10 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 11 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java (hunk 17 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeTemplate.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverter.java (hunk 19 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverter.java]: Unix timestamp in seconds :)
UnixTimeStampToDate

File: src/main/java/org/webtree/social/stackexchange/domain/User.java (hunk 21 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/domain/User.java:124]: Done.

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/UserTemplateIntegrationTest.java (hunk 27 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/integration/UserTemplateIntegrationTest.java]: Done.

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/SiteTemplateIntegrationTest.java (hunk 26 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/integration/SiteTemplateIntegrationTest.java]: Done.

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java (hunk 24 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java]: Done.

File: src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java (hunk 25 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/integration/AbstractApiIntegrationTest.java]: Done.

File: src/test/java/org/webtree/social/stackexchange/api/impl/UserTemplateTest.java (hunk 23 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/UserTemplateTest.java]: Done

File: src/main/java/org/webtree/social/stackexchange/api/StackExchange.java (hunk 2 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/StackExchange.java]: Done.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java (hunk 4 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: Renamed

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java (hunk 7 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: fixed

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java (hunk 5 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: fixed

File: src/main/java/org/webtree/social/stackexchange/domain/NetworkUser.java (hunk 20 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/domain/NetworkUser.java]: Done/

File: src/test/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandlerTest.java (hunk 22 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandlerTest.java:27]: Done.

File: src/test/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverterTest.java (hunk 28 above)
Comment (UdjinSkobelev) [reply src/test/java/org/webtree/social/stackexchange/converter/UnixTimestampToMillisecondsConverterTest.java]: Done

File: pom.xml [hunk 29]
```diff
@@ -20,46 +20,80 @@
...
-            <artifactId>spring-social-facebook</artifactId>
-            <version>2.0.3.RELEASE</version>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>26.0-jre</version>
+        </dependency>
+        <!--Test dependencies-->
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
```
Comment (Max-Levitskiy) [inline pom.xml:46]: Why do you need spring boot here?

File: src/main/java/org/webtree/social/stackexchange/api/ErrorCodes.java [hunk 30]
```diff
@@ -5,6 +5,7 @@
  */
 
 public class ErrorCodes {
+    public static final int UNKNOWN_ERROR = 0;
     public static final int BAD_PARAMETER = 400;
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/ErrorCodes.java]: You have http codes constants here: java.net.HttpURLConnection.

File: src/main/java/org/webtree/social/stackexchange/api/StackExchange.java [hunk 31]
```diff
@@ -3,7 +3,8 @@
 /**
  * Created by Udjin Skobelev on 27.09.2018.
  */
-public interface StackExchange extends Api{
+public interface StackExchange {
     UserOperations userOperations();
     SiteOperations siteOperations();
+    String getBaseApiUrl();
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/StackExchange.java]: why do you have this method on high level interface? I see usage only in test and don't see a reason for have it here.

File: src/main/java/org/webtree/social/stackexchange/api/StackObjectFetcher.java [hunk 32]
```diff
@@ -7,8 +7,7 @@
  * Created by Udjin Skobelev on 27.09.2018.
  */
 
-public interface Api {
+public interface StackObjectFetcher {
     <T> ResponseWrapper<T> fetchObject(String methodName, Class<T> itemsType);
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/StackObjectFetcher.java]: fetchObject -> fetch

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 33]
```diff
@@ -24,88 +26,80 @@
...
+            .put(INTERNAL_ERROR, (error) -> {
+                throw new InternalServerErrorException(STACK_EXCHANGE_PROVIDER_ID, error.getErrorMessage());
+            })
+            .put(THROTTLE_VIOLATION, (error) -> {
+                throw new RateLimitExceededException(STACK_EXCHANGE_PROVIDER_ID);
+            })
+            .put(EMPORARITY_UNAVAILABLE, (error) -> {
+                throw new ServerException(STACK_EXCHANGE_PROVIDER_ID, error.getErrorMessage());
+            })
+            .put(UNKNOWN_ERROR, (error) -> {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: It's wrong logic. You should make default handling in handleError() if this map does not contain error code. Now you missed situation with default case (new error code will fail with null pointer).

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 34]
```diff
@@ -24,88 +26,80 @@
...
-        while (reader.ready()) {
-            sb.append(reader.readLine());
+            return error;
+        } catch (JsonParseException exception) {
+            return new StackExchangeError(0, "Something goes wrong", null);
         }
-        return sb.toString();
     }
 }
+
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: Remove empty lines

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java [hunk 35]
```diff
@@ -24,88 +26,80 @@
...
+            logger.debug("   NAME     : {}", error.getErrorName());
 
-    private String readFully(InputStream in) throws IOException {
-        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
-        StringBuilder sb = new StringBuilder();
-        while (reader.ready()) {
-            sb.append(reader.readLine());
+            return error;
+        } catch (JsonParseException exception) {
+            return new StackExchangeError(0, "Something goes wrong", null);
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: It's yours exception. Don't put nulls and magic numbers there. 
And it's not "unknown" exception. When you re-throw exception, you shouldn't lose any information.
Here you have JsonParseException. Only its name told a lot. Also it has two properties.

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+        this.apiUrl = apiUrl;
+        this.restTemplate = restTemplate;
+    }
+
+    @Override
+    public <T> ResponseWrapper<T> fetchObject(String method, Class<T> type) {
+        URI uri = URIBuilder.fromUri(apiUrl + method).build();
+        return restTemplate.exchange(uri, HttpMethod.GET, null, new ParameterizedTypeReference<ResponseWrapper<T>>() {
+            public Type getType() {
+                return new ParameterizedTypeImpl((ParameterizedType) super.getType(), new Type[]{type});
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java:33]: Why do you implement this? Oo 
For get correct type you just need to have this 
`new ParameterizedTypeReference<ResponseWrapper<T>>(){}`

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.net.URI;
+
+/**
+ * Created by Udjin Skobelev on 10.10.2018.
+ */
+
+public class StackObjectFetcherImpl implements StackObjectFetcher {
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java:19]: Missed test

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        }).getBody();
+    }
+
+    @Override
+    public <T> ResponseWrapper<T> fetchObject(String methodName, Class<T> itemsType, MultiValueMap<String, String> queryParams) {
+        URI uri = URIBuilder.fromUri(apiUrl + methodName).queryParams(queryParams).build();
+        return restTemplate.exchange(uri, HttpMethod.GET, null, new ParameterizedTypeReference<ResponseWrapper<T>>() {
+            public Type getType() {
+                return new ParameterizedTypeImpl((ParameterizedType) super.getType(), new Type[]{itemsType});
```
Comment (Max-Levitskiy) [inline src/main/java/org/webtree/social/stackexchange/api/impl/StackObjectFetcherImpl.java:43]: same here

File: pom.xml [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>2.22.0</version>
+                <configuration>
+                    <excludes>
+                        <exclude>**/integration/*.java</exclude>
```
Comment (Max-Levitskiy) [inline pom.xml]: Move integration tests in separate package. It shouldn't be excluded by stars. How are you gonna run just units from IDE? :)

**Review by Max-Levitskiy — CHANGES_REQUESTED** [review]

File: src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java (hunk 34 above)
Comment (UdjinSkobelev) [reply src/main/java/org/webtree/social/stackexchange/api/impl/StackExchangeErrorHandler.java]: Done

**Pull request merged**